libhb/decavcodec.c

```diff
@@ -1515,11 +1565,58 @@ static int decavcodecvInit( hb_work_object_t * w, hb_job_t * job )
         return 1;
     }
 
+#if HB_PROJECT_FEATURE_NVENC
```

Comment:
This should be moved to work.c, where the pipeline is configured and the choice of which pixel format to use is made.
It should try to use the hardware decoder even when scanning, so it's possible to know if it actually can decode the video stream or not, and then provide an option to enable or disable the hardware decoder in the hb_job_t struct, decavcodec.c should just use the currently options set in the job.
So the hardware encoder really need to be able to copy the frames to the main cpu memory, even if only for the scan pass.

Reply:
>It should try to use the hardware decoder even when scanning

That will cause not only HW frame download but sampling conversion as well because native format for Nvdec is NV12 while Handbrake expects YUV420.

Reply:
>This should be moved to work.c, where the pipeline is configured and the choice of which pixel format to use is made

Where does this happen? 
I went through the libhb/work.c searching for the `HB_PROJECT_FEATURE_QSV` entries, there are as follows:

1. `work_func()`
2. `hb_display_job_info()`
3. `sanitize_filter_list()`
4. `do_job()`
5. `filter_loop()`

But scan pass happens before any of those are called.

Reply:
I didn't check if QSV is actually used during a scan, but the pipeline format is set in work.c by calling hb_get_best_pix_fmt(), so at least that part should be moved there. This requires a bit of thinking on where to place every check.
Eventually decavcodec.c should only check if there is a job (if we don't want to use the hw decoder for the scan phase) and if the hw decoder is enabled in the job.

The hw_device_ctx in hb_title_t should be moved the the hb_job_t, if there isn't a reason to keep it in the title struct, the hb_title_t should be an immutable struct after the scan is done.

Reply:
Hi @galad87 @sr55 

Please check out ToT, it addresses the issues you've mentioned.

Reply:
It's still modifying hb_title_t in work.c. hb_title_t should only be modified during a scan, and then you need a hb_job_t filed to enable or disable nvdec.

Reply:
I've pushed a patch that should address this.